autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/ScalingRequest.java

```diff
@@ -56,21 +60,27 @@
     @Inject
     private PeriscopeMetricService metricService;
 
-    public ScalingRequest(Cluster cluster, ScalingPolicy policy, int totalNodes, int desiredNodeCount) {
+    @Inject
+    private CloudbreakCommunicator cloudbreakCommunicator;
+
+    public ScalingRequest(Cluster cluster, ScalingPolicy policy, int totalNodes, int desiredNodeCount, List<String> decommissionNodeIds) {
         this.cluster = cluster;
         this.policy = policy;
         this.totalNodes = totalNodes;
         this.desiredNodeCount = desiredNodeCount;
+        this.decommissionNodeIds = decommissionNodeIds;
     }
 
     @Override
     public void run() {
         MDCBuilder.buildMdcContext(cluster);
         try {
             int scalingAdjustment = desiredNodeCount - totalNodes;
-            if (scalingAdjustment > 0) {
+            if (!decommissionNodeIds.isEmpty()) {
+                scaleDownByNodeIds(decommissionNodeIds);
+            } else if (scalingAdjustment > 0) {
                 scaleUp(scalingAdjustment, totalNodes);
-            } else {
+            } else if (scalingAdjustment < 0) {
```

Comment:
This will be triggered in case of TimeAlerts. Will eventually have to change that over to work with the YARN API as well - so that nodes are not removed randomly. Separate jira / patch.

Reply:
Also, I suspect this gets triggered if the cluster has somehow managed to go above the maximum size. e.g. the max node count is reconfigured.

Reply:
This is basically for TimeAlerts.

Reply:
Also for the change in 'max-nodes' scenario?

Reply:
I purposefully did not downscale when the cluster has somehow gone above maximum size which was caught in my unit testing, because in this case the user might have scale up cluster manually and if autoscale reduces the node size it will override the user which i felt is not expected behaviour of autoscaling.

Reply:
That is an interesting point. Think we need a little more discussion on what the behaviour should be here.
As an end user, if I scale up a cluster manually beyond maxNodes
1) If there is no load on the cluster, it will be scaled down immediately - nullifying the scale up. (e.g. scale up manually to submit a large job)
2) If there a load on the cluster already, it will not be scaled down to maxNodes.
I think it is fair to keep the cluster between minNodes and maxNodes if AutoScaling is enabled. If a user wants to change something - disable the policy and scale to whatever size is needed.